src/lib/operators/join_nested_loop.cpp

```diff
@@ -0,0 +1,239 @@
+                               const std::shared_ptr<const AbstractOperator> right, const JoinMode mode,
+                               const std::pair<ColumnID, ColumnID>& column_ids, const ScanType scan_type)
+    : AbstractJoinOperator(left, right, mode, column_ids, scan_type) {}
+
+const std::string JoinNestedLoop::name() const { return "JoinNestedLoop"; }
+
+std::shared_ptr<AbstractOperator> JoinNestedLoop::recreate(const std::vector<AllParameterVariant>& args) const {
+  return std::make_shared<JoinNestedLoop>(_input_left->recreate(args), _input_right->recreate(args), _mode, _column_ids,
+                                          _scan_type);
+}
+
+std::shared_ptr<const Table> JoinNestedLoop::_on_execute() {
+  PerformanceWarning("Nested Loop Join used");
+
+  _create_table_structure();
+
+  _left_column_id = _column_ids.first;
+  _right_column_id = _column_ids.second;
+
+  // Resolve both column data types and execute the main join loop
+  resolve_data_type(_left_in_table->column_type(_left_column_id), [&](auto left_type) {
+    using LeftType = typename decltype(left_type)::type;
+
+    resolve_data_type(_right_in_table->column_type(_right_column_id), [&](auto right_type) {
+      using RightType = typename decltype(right_type)::type;
+
+      // for the right outer join we switch types because we treat the "left" side as the outer relation
+      if (_mode == JoinMode::Right) {
+        this->_perform_join<RightType, LeftType>();
+      } else {
+        this->_perform_join<LeftType, RightType>();
+      }
+    });
+  });
+
+  return _output_table;
+}
+
+void JoinNestedLoop::_create_table_structure() {
+  _output_table = std::make_shared<Table>();
+
+  _left_in_table = _input_left->get_output();
+  _right_in_table = _input_right->get_output();
+
+  const bool left_may_produce_null = (_mode == JoinMode::Right || _mode == JoinMode::Outer);
+  const bool right_may_produce_null = (_mode == JoinMode::Left || _mode == JoinMode::Outer);
+
+  // Preparing output table by adding columns from left table
+  for (ColumnID column_id{0}; column_id < _left_in_table->column_count(); ++column_id) {
+    auto nullable = (left_may_produce_null || _left_in_table->column_is_nullable(column_id));
+    _output_table->add_column_definition(_left_in_table->column_name(column_id), _left_in_table->column_type(column_id),
+                                         nullable);
+  }
+
+  // Preparing output table by adding columns from right table
+  for (ColumnID column_id{0}; column_id < _right_in_table->column_count(); ++column_id) {
+    auto nullable = (right_may_produce_null || _right_in_table->column_is_nullable(column_id));
+    _output_table->add_column_definition(_right_in_table->column_name(column_id),
+                                         _right_in_table->column_type(column_id), nullable);
+  }
+}
+
+template <typename LeftType, typename RightType>
+void JoinNestedLoop::_perform_join() {
```

Comment:
The method gets compiled 25 times at the moment, when it should be 17 (4 numerical types ^ 2 + strings). In `column_comparison_table_scan_impl.cpp` I use a constexpr-if to compare the types and filter out unwanted combinations.

Reply:
Your point makes perfect sense. In order to keep this as a reaaally short example on how to do a simple join and how to use iterables, I would leave it like it is.